controllers/databasecluster_controller.go

```diff
@@ -2141,18 +2203,48 @@ func (r *DatabaseClusterReconciler) genPGDataSourceSpec(ctx context.Context, dat
 
 		backupBaseName = filepath.Base(*dbClusterBackup.Status.Destination)
 		backupStorageName = dbClusterBackup.Spec.BackupStorageName
+		dest = *dbClusterBackup.Status.Destination
 	}
 
 	if database.Spec.DataSource.BackupSource != nil {
 		backupBaseName = filepath.Base(database.Spec.DataSource.BackupSource.Path)
 		backupStorageName = database.Spec.DataSource.BackupSource.BackupStorageName
 	}
 
+	backupStorage := &everestv1alpha1.BackupStorage{}
+	err := r.Get(ctx, types.NamespacedName{Name: backupStorageName, Namespace: r.defaultNamespace}, backupStorage)
+	if err != nil {
+		return nil, errors.Join(err, fmt.Errorf("failed to get backup storage %s", backupStorageName))
+	}
+	if database.Namespace != r.defaultNamespace {
+		if err := r.reconcileBackupStorageSecret(ctx, backupStorage, database); err != nil {
+			return nil, err
+		}
+	}
+
+	if dest == "" {
+		dest = "/" + backupStoragePrefix(database)
+	} else {
+		// Extract the relevant prefix from the backup destination
+		switch backupStorage.Spec.Type {
+		case everestv1alpha1.BackupStorageTypeS3:
+			dest = strings.TrimPrefix(dest, "s3://")
+		case everestv1alpha1.BackupStorageTypeAzure:
+			dest = strings.TrimPrefix(dest, "azure://")
+		}
+
+		dest = strings.TrimPrefix(dest, backupStorage.Spec.Bucket)
+		dest = strings.TrimLeft(dest, "/")
+		prefix := backupStoragePrefix(database)
+		prefixCount := len(strings.Split(prefix, "/"))
+		dest = "/" + strings.Join(strings.SplitN(dest, "/", prefixCount+1)[0:prefixCount], "/")
```

Comment:
It would be great if we could cover this transformation logic with a unit test that would also document the expected output

Reply:
Awesome, thanks!